appinventor/appengine/src/com/google/appinventor/server/project/youngandroid/YoungAndroidProjectService.java

```diff
@@ -911,4 +1046,94 @@ public RpcResult getBuildResult(User user, long projectId, String target) {
   public int getCurrentProgress(User user, long projectId, String target) {
     return storageIo.getBuildStatus(user.getUserId(), projectId);
   }
+
+  /**
+   * This method reads from a stream based on the passed connection. It reads
+   * the content as bytes, so it can deal with binary files
+   *
+   * @param connection the connection to read from
+   * @return the contents of the stream
+   * @throws IOException if it cannot read from the http connection
+   */
+  private static byte[] getResponseBytes(HttpURLConnection connection) throws IOException {
+    // Use the content encoding to convert bytes to characters.
+    InputStream input = connection.getInputStream();
+    int bytesRead = 0;
+    int contentLength = connection.getContentLength();
+    LOG.info("contentLength = " + contentLength);
+    byte buffer[] = new byte[contentLength];
+    while (true) {
+      int i = input.read(buffer, bytesRead, contentLength - bytesRead);
+      if (i < 0) {
+        break;
+      }
+      bytesRead += i;
+    }
+    LOG.info("Done, contentLenght = " + contentLength + " bytesRead = " + bytesRead);
+    return buffer;
+  }
+
+  /*
+   * Verify that the input projectName is unique among the user's
+   * projects.
+   */
+
+  private String verifyProjectName(String userId, String projectName) {
+    int count = 0;
+    List<Long> projectIds = storageIo.getProjects(userId);
+    List<UserProject> projects = storageIo.getUserProjects(userId, projectIds);
+    List<String> projectNames = new ArrayList();
```

Comment:
A minor improvement would be to use a TreeSet and request an iterator starting at the point where the project name collides and walk the (sorted) project names from there. Granted, most people will only have a few projects max (average is ~5 per user).

Reply:
Done. I didn't change the code, I just swapped out the ArrayList for a TreeSet. The TreeSet (well, really TreeMap) implementation of "contains" does binary search, which is more efficient. There is overhead when the TreeSet is created, but that is done once while the "contains" may be called multiple times if we have multiple collisions. Note: New code tested, complete with forced collisions.